### ISSUE TITLE
Potential bugs in node gRPC handler for receiving tasks and sending replies

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#1668]
I've been investigating the issue that comes up from time to time that is `Received reply from unexpected request request_id:'....'`. 

- This issue occurs if the same node is launched twice or more. Both node manages the connect to server and Requests module confuses who to reply or receive answer, therefore requests IDs are mixed up. 

However, while addressing this issue following potential bugs are identified by coding agent in the `transport/client.py`:

# Bug Report: Multiple Critical Issues in gRPC Client Implementation

## Summary
Several critical bugs and potential issues have been identified in `fedbiomed/transport/client.py` that could lead to runtime errors, resource leaks, and incorrect error handling.

## Environment
- **File**: `fedbiomed/transport/client.py`
- **Component**: gRPC Client (Node-Researcher Communication)

---

### 1. Race Condition: Double `task_done()` Call in Sender
**Severity**: High  
**Location**: `Sender._handle_after_process()` (~line 433)

**Description**:
`queue.task_done()` is called unconditionally in the else branch, which can be triggered multiple times if retry count exceeds `MAX_SEND_RETRIES`. This causes:

*Current Code**:
```python
else:
    if self._retry_count >= MAX_SEND_RETRIES:
        logger.warning(f"Message can not be sent to researcher after {MAX_SEND_RETRIES} retries...")
    # Only cleanup if not already done (defensive against double task_done)
    self._queue.task_done()
    self._retry_count = 0
    self._retry_item = None
```

Proposed fix:
 
```
else:
    if self._retry_count >= MAX_SEND_RETRIES:
        logger.warning(f"Message can not be sent to researcher after {MAX_SEND_RETRIES} retries...")
    
    # Actual defensive check
    if self._retry_item is not None:
        self._queue.task_done()
        self._retry_item = None
        self._retry_count = 0
        self._stub_type = _StubType.NO_STUB
```

### 2. Missing Retry State Cleanup After Successful Send
Severity: High
Location: Sender._call_researcher() (~lines 462-490)

Description:
After successful message transmission, retry state variables are not reset. This means:

Subsequent failures may reference stale _retry_item
task_done() may not be called, blocking the queue
Retry counter persists incorrectly
Current Code:

```
if isinstance(stub_function, grpc.aio.UnaryUnaryMultiCallable):
    await stub_function(item["message"].to_proto())
    # No cleanup - success path doesn't reset state

elif isinstance(stub_function, grpc.aio.StreamUnaryMultiCallable):
    # ...
    await stream_call.done_writing()
    # No cleanup here either
```

Proposed fix 

```
if isinstance(stub_function, grpc.aio.UnaryUnaryMultiCallable):
    await stub_function(item["message"].to_proto())
    # Cleanup on success
    self._queue.task_done()
    self._retry_count = 0
    self._retry_item = None
    self._stub_type = _StubType.NO_STUB

elif isinstance(stub_function, grpc.aio.StreamUnaryMultiCallable):
    stream_call = stub_function()
    for reply in self._stream_reply(item["message"]):
        await stream_call.write(reply)
    await stream_call.done_writing()
    
    # Cleanup on success
    self._queue.task_done()
    self._retry_count = 0
    self._retry_item = Non
```

### 3.  Incorrect Exception Handling: Catching

Severity: High
Location: Listener._listen() (~line 216)

Description:
GeneratorExit should never be caught as it's used by Python's runtime to clean up generators/coroutines. Catching it prevents proper cleanup and can cause undefined behavior.

Current code: 
```
except (Exception, GeneratorExit) as exp:
    logger.error(f"Unexpected error raised by node gRPC client...")
```

Proposed fix: 

```
except Exception as exp:
    logger.error(f"Unexpected error raised by node gRPC client...")
```

### 4.  Memory Inefficiency: O(n²) Bytes Concatenation

Severity: Medium
Location: TaskListener._call_researcher() (~line 353)

Description:
Concatenating bytes in a loop (reply += answer.bytes_) creates new byte objects on each iteration, leading to quadratic memory usage for large payloads.

Current code: 
```
reply = bytes()
async for answer in iterator:
    reply += answer.bytes_
    if answer.size != answer.iteration:
        continue
```

Proposed fix: 

```
reply_chunks = []
async for answer in iterator:
    reply_chunks.append(answer.bytes_)
    if answer.size != answer.iteration:
        continue
    else:
        reply = b''.join(reply_chunks)
        # ... process reply
        reply_chunks.clear()
```

### 5. Missing Timeout on Channel Close

Severity: Medium
Location: Channels.connect() (~line 92)

Description:
channel.close() can hang indefinitely if there are pending operations or network issues.

Current Code:

```python
for st, channel in self._channels.items():
    if channel and (stub_type == _StubType.ANY_STUB or stub_type == st):
        await channel.close()
```

Proposed fix: 

```
for st, channel in self._channels.items():
    if channel and (stub_type == _StubType.ANY_STUB or stub_type == st):
        try:
            await asyncio.wait_for(channel.close(), timeout=5.0)
        except asyncio.TimeoutError:
            logger.warning(f"Channel close timeout for {st}, forcing closure"
```

### 6. Unhandled SSL Certificate Retrieval Errors

Severity: Medium
Location: GrpcClient._connect() (~line 270)

Description:
ssl.get_server_certificate() can raise various exceptions (connection errors, SSL errors, timeout) that aren't caught, potentially causing the connection loop to crash.

Current Code:
```python
self._researcher.certificate = bytes(
    ssl.get_server_certificate(
        (self._researcher.host, self._researcher.port)
    ),
    "utf-8",
)
```

proposed fix: 

```
try:
    self._researcher.certificate = bytes(
        ssl.get_server_certificate(
            (self._researcher.host, self._researcher.port)
        ),
        "utf-8",
    )
except (ssl.SSLError, OSError) as e:
    logger.warning(f"Failed to retrieve server certificate: {e}")
    await asyncio.sleep(GRPC_CLIENT_CONN_RETRY_TIMEOUT)
    continue
```

## Reproduction steps: 

```
Issue #1 & #2: Run node with unreliable network, trigger retry exhaustion
Issue #3: Close a running gRPC client task externally
Issue #4: Send large payloads (>100MB) and monitor memory usage
Issue #5: Kill researcher server during node channel close
Issue #6: Point node to invalid/inaccessible server
```
